python_modules/dagster/dagster/_core/definitions/freshness_checks.py

```diff
@@ -0,0 +1,257 @@
+from dagster._core.event_api import AssetRecordsFilter, EventLogRecord
+from dagster._core.events import DagsterEventType
+from dagster._core.execution.context.compute import AssetExecutionContext
+from dagster._core.instance import DagsterInstance
+from dagster._utils.schedules import is_valid_cron_string, reverse_cron_string_iterator
+
+from .asset_check_result import AssetCheckResult
+from .asset_checks import AssetChecksDefinition
+from .assets import AssetsDefinition, SourceAsset
+from .events import AssetKey, CoercibleToAssetKey
+from .partition import PartitionsDefinition
+from .time_window_partitions import TimeWindowPartitionsDefinition
+
+DEFAULT_FRESHNESS_SEVERITY = AssetCheckSeverity.WARN
+
+
+def build_freshness_checks(
+    assets: Sequence[Union[CoercibleToAssetKey, AssetsDefinition, SourceAsset]],
+    freshness_cron: Optional[str] = None,
+    maximum_lag_minutes: Optional[int] = None,
+    severity: Optional[AssetCheckSeverity] = None,
+) -> Sequence[AssetChecksDefinition]:
+    """For each provided asset, constructs a freshness check definition.
+
+    An asset is considered fresh if it has been materialized or observed within a certain time window.
+    The freshness_cron and maximum_lag_minutes parameters are used to define this time window.
+    freshness_cron defines the time at which the asset should have been materialized, and maximum_lag_minutes provides a tolerance for the range at which the asset can arrive.
+
+    Let's say an asset kicks off materializing at 12:00 PM and takes 10 minutes to complete.
+    Allowing for operational constraints and delays, the asset should always be materialized by 12:30 PM.
+    12:00 PM provides the lower boundary, since the asset kicks off no earlier than this time.
+    Then, we set freshness_cron to "30 12 * * *", which means the asset is expected by 12:30 PM, and maximum_lag_minutes to 30,
+    which means the asset can be materialized no earlier than 12:00 PM.
+
+    A time-window partitioned asset is considered fresh if the latest completed time window is materialized or observed.
+    The freshness_cron parameter can be used to influence the time window that is checked. For example, if I have a daily partitioned asset,
+    and I specify a freshness_cron for 12:00 PM, and I run the check at 1:00 AM March 12, I will check for the time window
+    March 10 12:00 AM - March 11 12:00 AM, since the most recent completed tick of the cron was at 12:00 PM March 11.
+    For partitioned assets, the maximum_lag_minutes parameter is not used.
+
+    Args:
+        assets (Sequence[Union[CoercibleToAssetKey, AssetsDefinition, SourceAsset]): The assets to construct checks for. For each passed in
+            asset, there will be a corresponding constructed `AssetChecksDefinition`.
+        freshness_cron (Optional[str]): The cron for which freshness is defined.
+        maximum_lag_minutes (Optional[int]): The maximum lag in minutes that is acceptable for the asset.
```

Comment:
Similar to the suggestion above: is there some way of phrasing this in terms of "The check will pass if"?